src/core/sock.c

```diff
@@ -824,11 +824,13 @@ static void nn_sock_shutdown (struct nn_fsm *self, int src, int type,
     }
     if (nn_slow (sock->state == NN_SOCK_STATE_STOPPING_EPS)) {
 
+        /* We are no longer interested in pipe events */
+        if (src == 0)
+            return;
```

Comment:
I'm looking at this, and I think maybe we miss the cleanup that would occur in finish2.  Probably instead of return here, we should just go to finish2.